optimizely/config_manager.py

```diff
@@ -0,0 +1,198 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import http
+import requests
+import threading
+import time
+from requests import exceptions as requests_exceptions
+
+from optimizely import exceptions
+from optimizely import logger as _logging
+from optimizely import project_config
+from optimizely.error_handler import NoOpErrorHandler as noop_error_handler
+from optimizely.helpers import enums
+
+
+class BaseConfigManager(object):
+  """ Base class for Optimizely's config manager. """
+
+  def get_config(self):
+    """ Get config for use by optimizely.Optimizely.
+    The config should be an instance of project_config.ProjectConfig."""
+    raise NotImplementedError
+
+
+class StaticConfigManager(BaseConfigManager):
+  """ Config manager that returns ProjectConfig based on provided datafile. """
+
+  def __init__(self,
+               datafile,
+               logger=None,
+               error_handler=None):
+    self.logger = logger or _logging.adapt_logger(logger or _logging.NoOpLogger())
+    self.error_handler = error_handler or noop_error_handler
+    self._datafile = datafile
+    self._config = project_config.ProjectConfig(self._datafile, self.logger, self.error_handler)
+
+  def get_config(self):
+    """ Returns instance of ProjectConfig.
+
+    Returns:
+      ProjectConfig.
+    """
+    return self._config
+
+
+class PollingConfigManager(BaseConfigManager):
+  """ Config manager that polls for the datafile and updated ProjectConfig based on an update interval. """
+
+  def __init__(self,
+               sdk_key=None,
+               update_interval=None,
+               url=None,
+               url_template=None,
+               logger=None,
+               error_handler=None):
+    """ Initialize config manager. One of sdk_key or url has to be set to be able to use.
+
+    Args:
+      sdk_key: Optional string uniquely identifying the datafile.
+      update_interval: Optional floating point number representing time interval in seconds
+                       at which to request datafile and set ProjectConfig.
+      url: Optional string representing URL from where to fetch the datafile. If set it supersedes the sdk_key.
+      url_template: Optional string template which in conjunction with sdk_key
+                    determines URL from where to fetch the datafile.
+      logger: Provides a logger instance.
+      error_handler: Provides a handle_error method to handle exceptions.
+    """
+    self.datafile_url = self.get_datafile_url(sdk_key, url, url_template)
+    self.update_interval = self.get_update_interval(update_interval)
+    self.last_modified = None
+    self._datafile = None
+    self._config = None
+    self._polling_thread = threading.Thread(target=self._run)
+    self.is_running = False
```

Comment:
Instead of managing our own `is_running`, should we delegate that to `self._polling_thread.isAlive()`? 

https://docs.python.org/3/library/threading.html#threading.Thread.is_alive

This class could expose an `@property` for that:

```
@property
def is_running(self):
  return self._polling_thread.isAlive()
```

I believe the semantics are the same, but worth double-checking my understanding.

Reply:
Nice. Will take a look.